src/main/java/org/candlepin/subscriptions/tally/roller/DailySnapshotRoller.java

```diff
@@ -0,0 +1,136 @@
+            //       calculators to a list and applying the host to each.
+            String accountNumber = hostEntry.getKey();
+            ProductUsageCalculation rhelUsageCalculation = new ProductUsageCalculation(accountNumber,
+                this.product);
+            hostEntry.getValue().forEach(host -> {
+                NormalizedFacts facts = factNormalizer.normalize(host);
+                if (facts.getProducts().contains(rhelUsageCalculation.getProductId())) {
+                    rhelUsageCalculation.addCores(facts.getCores() != null ? facts.getCores() : 0);
+                    rhelUsageCalculation.addInstance();
+                }
+
+                // Validate and set the owner.
+                String owner = facts.getOwner();
+                if (owner == null) {
+                    // Don't set null owner as it may overwrite an existing value.
+                    // Likely won't happen, but there could be stale data in inventory
+                    // with no owner set.
+                    return;
+                }
+
+                String currentOwner = rhelUsageCalculation.getOwner();
+                if (currentOwner != null && !currentOwner.equalsIgnoreCase(owner)) {
+                    throw new IllegalStateException(
+                        String.format("Attempt to set a different owner for an account: %s:%s",
+                            currentOwner, owner));
+                }
+                rhelUsageCalculation.setOwner(owner);
+            });
+            log.debug("Found {} cores of RHEL for account: {}", rhelUsageCalculation.getTotalCores(),
+                rhelUsageCalculation.getAccount());
+            calculations.add(rhelUsageCalculation);
+        }
+        return calculations;
+    }
+
+    private void produceDailySnapshots(List<String> accounts, List<ProductUsageCalculation> calculations) {
+
+        Map<String, AccountMaxValues> maxValues = calculations.stream()
+            .map(this::createMaxValuesFromCalculation)
+            .collect(Collectors.toMap(AccountMaxValues::getAccountNumber, Function.identity()));
+
+        log.debug("Fetching existing daily snapshots for {} accounts.", accounts.size());
+
+        Map<String, TallySnapshot> existingRhelSnapsForToday =
+            tallyRepo.findByAccountNumberInAndProductIdAndGranularityAndSnapshotDateBetween(
+            accounts, this.product, TallyGranularity.DAILY, clock.startOfToday(), clock.endOfToday())
+            .collect(Collectors.toMap(TallySnapshot::getAccountNumber, Function.identity()));
+
+        log.debug("Found {} existing snapshots for today.", existingRhelSnapsForToday.size());
+
+        updateSnapshots(existingRhelSnapsForToday, maxValues, TallyGranularity.DAILY);
+    }
+
+    private AccountMaxValues createMaxValuesFromCalculation(ProductUsageCalculation calculation) {
+        return new AccountMaxValues(calculation.getAccount(), calculation.getOwner(),
```

Comment:
Would it make sense just to add a convenience constructor to AccountMaxValues that takes a ProductUsageCalculation object?

Reply:
DONE